gips/0046-graphql-validations.md

```diff
@@ -0,0 +1,108 @@
+- `NoFragmentsCycle` (https://spec.graphql.org/draft/#sec-Fragment-spreads-must-not-form-cycles)
+- `PossibleFragmentSpreads` (https://spec.graphql.org/draft/#sec-Fragment-spread-is-possible)
+- `NoUnusedVariables` (https://spec.graphql.org/draft/#sec-All-Variables-Used)
+- `NoUndefinedVariables` (https://spec.graphql.org/draft/#sec-All-Variable-Uses-Defined)
+- `KnownArgumentNames` (https://spec.graphql.org/draft/#sec-Argument-Names , https://spec.graphql.org/draft/#sec-Directives-Are-In-Valid-Locations)
+- `UniqueArgumentNames` (https://spec.graphql.org/draft/#sec-Argument-Names)
+- `UniqueVariableNames` (https://spec.graphql.org/draft/#sec-Variable-Uniqueness)
+- `ProvidedRequiredArguments` (https://spec.graphql.org/draft/#sec-Required-Arguments)
+- `ValuesOfCorrectType` (https://spec.graphql.org/draft/#sec-Values-of-Correct-Type)
+- `VariablesInAllowedPosition` (https://spec.graphql.org/draft/#sec-All-Variable-Usages-Are-Allowed)
+- `KnownDirectives` (https://spec.graphql.org/draft/#sec-Directives-Are-Defined)
+- `UniqueDirectivesPerLocation` (https://spec.graphql.org/draft/#sec-Directives-Are-Unique-per-Location)
+
+The following spec rules are not part of this implementation:
+
+- `ExecutableDefinitions` - not needed, because of `graphql_parser` nature.
+- `UniqueInputFieldNames` (blocked by https://github.com/graphql-rust/graphql-parser/issues/59, low priority for `graph-node`)
+
+In case of a validation error as part of the `validate` phase, `graph-node` will return a validation error, and will skip the GraphQL execution:
+
+![GraphQL Validation Error](../assets/gip-0046/graphql-validations-1.png)
+
+![GraphQL Validation Error](../assets/gip-0046/graphql-validations-2.png)
+
+# Backwards Compatibility
+
+The GraphQL execution layer will run in the same way, but due to the nature of the extra added validation checks, some users might experience a change in the way `graph-node` responds to end-users GraphQL operations.
+
+To address the potential breaking change for users, we applied the following measures:
+
+1. Document the change, how it might effect consumers and how to migrate/fix invalid operations (https://github.com/graphprotocol/docs/pull/284)
+2. Provide a CLI tool that can locate potential operations that might be invalid at runtime (documented in https://github.com/graphprotocol/docs/pull/284)
+3. Find high-traffic and high-profile Subgraphs and contact the owners, in order to give them enough time to fix invalid operations.
+4. Run GraphQL validations as part of the Studio GraphiQL editor, to make sure users are aware of this change ahead of time (https://github.com/edgeandnode/graphiql-playground/pull/17).
+5. Announce GraphQL validations is coming, and provide enough time for consumers to adjust.
+
+Due to the nature of this change, only end-users and not Subgraph developers are effected. We are not able to find all effected users, but we can give users enough time to adjust and fix their invalid GraphQL operations.
```

Comment:
I think end-users and Subgraph developers are slightly vague.
"This change will affect developers querying subgraphs, not developers solely deploying and publishing subgraphs. We are not able to definitively identify all affected users, but we can give users enough time to adjust and fix their invalid GraphQL operations.".

Reply:
yeah that's much better. thanks @azf20